modal/_container_entrypoint.py

```diff
@@ -236,9 +240,18 @@ async def _send_outputs(self):
             # TODO(erikbern): we'll get a RESOURCE_EXCHAUSTED if the buffer is full server-side.
             # It's possible we want to retry "harder" for this particular error.
 
-    async def run_inputs_outputs(self):
+    async def _acquire_semaphore(self):
+        await self.semaphore.put(None)
+
+    async def _release_semaphore(self):
+        await self.semaphore.get()
+        self.semaphore.task_done()
+
+    async def run_inputs_outputs(self, input_concurrency: int = 1):
         # This also makes sure to terminate the outputs
         self.output_queue: asyncio.Queue = asyncio.Queue()
+        # Ensure we do not fetch new inputs when container is too busy
+        self.semaphore: asyncio.Queue[None] = asyncio.Queue(maxsize=input_concurrency)
```

Comment:
also this should be even higher up since it's the place where it's being defined

Reply:
actually maybe `self.semaphore` should be initialized in the constructor?

Reply:
wait why is self.semaphore a queue? I just assumed it was an `asyncio.Queue` object based on the name... let's rename the variable or use an actual semaphore

Reply:
ok i'll rename the variable. 

when we receive EOF, we need to wait for all the outputs to get enqueued. we use `await semaphore.join()` to know when that happens.

Reply:
this makes it feel like we should use a thread pool here?

Reply:
I refactored to use a semaphore. The "wait for everything to be done" can be accomplished by collecting all the semaphore locks.

`for _ in range(n): await self._semaphore.acquire()`